### PR TITLE
Unpin commented out unsafe packages

### LINKED ISSUE #973: pip-compile bumps setuptools within commented-out spec
If you have a `requirements.in` which contains a package which depends on `setuptools`, for example:
```
$ cat requirements.in 
protobuf  # or google-api-core, ipython, pandas-gbq, pydata-google-auth
```
Then after running `pip-compile`, you get:
```
$ cat requirements.txt
#
# This file is autogenerated by pip-compile
# To update, run:
#
#    pip-compile
#
protobuf==3.10.0
six==1.12.0               # via protobuf

# The following packages are considered to be unsafe in a requirements file:
# setuptools==41.4.0        # via protobuf
```

As you can see, `pip-compile` refuses to include `setuptools` directly in a generated `requirements.txt`, instead adding a section with the comment "The following packages are considered to be unsafe in a requirements file:".
This behaviour is fine in the general (and I don't disagree with the comment or behaviour).

However because the commented out `setuptools` line includes a version number, when `setuptools` releases an update then the `requirements.txt` file ends up looking out of date.

Say you had committed a `requirements.txt` which looked like:
```
# requirements...

# The following packages are considered to be unsafe in a requirements file:
# setuptools==41.4.0        # via protobuf
```

Then `setuptools` releases `41.5.0` and now when you run `pip-compile` you get:
```
# requirements...

# The following packages are considered to be unsafe in a requirements file:
# setuptools==41.5.0        # via protobuf
```

This is inconvenient if you happen to regenerate your `requirements.txt` at some point after an upstream releases, however if you're also using `pip-compile` to check that your `requirements.txt` is in step with your `requirements.in` (see #882) in CI then it means that your build now fails for what is essentially a non-change.

I'm not sure what the right solution is here, though perhaps if there was a way for pip-compile to understand its own commented out dependency that might help?

### PR DESCRIPTION
**Pull request by atugushev**
Resolves #973. See the proposal in https://github.com/jazzband/pip-tools/issues/973#issuecomment-547010219.

**Changelog-friendly one-liner**: Unpin commented out unsafe packages in `requirements.txt`.

##### Contributor checklist

- [X] Provided the tests for the changes.
- [X] Gave a clear one-line description in the PR (that the maintainers can add to CHANGELOG.md on release).
- [x] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

**Comment by atugushev** [comment]
It turns out it was already discussed before, see #276 (comment), and fixed in #294, but has been broken later unintentionally.

File: tests/test_cli_compile.py [hunk 1]
```diff
@@ -691,8 +691,7 @@ def test_annotate_option(pip_conf, runner, option, expected):
 
 
 @pytest.mark.parametrize(
-    "option, expected",
-    [("--allow-unsafe", "\nsetuptools=="), (None, "\n# setuptools==")],
+    "option, expected", [("--allow-unsafe", "\nsetuptools=="), (None, "\n# setuptools")]
```
Comment (webknjaz) [inline tests/test_cli_compile.py]: I don't see any difference in this change. Looks unrelated...

File: piptools/writer.py [hunk 2]
```diff
@@ -189,17 +189,18 @@ def _iter_lines(
...
-                    marker=markers.get(key_from_ireq(ireq)),
-                    hashes=hashes,
-                )
+                ireq_key = key_from_ireq(ireq)
                 if not self.allow_unsafe:
-                    yield comment("# {}".format(req))
+                    yield comment("# {}".format(ireq_key))
                 else:
-                    yield req
+                    line = self._format_requirement(
```
Comment (webknjaz) [inline piptools/writer.py]: Maybe just `yield` it right away? I mean, it's expected that only lines are yielded by this generator function anyway, so this variable name doesn't add much semantical meaning to the code.

File: piptools/writer.py [hunk 3]
```diff
@@ -189,17 +189,18 @@ def _iter_lines(
...
 
             for ireq in unsafe_requirements:
-                req = self._format_requirement(
-                    ireq,
-                    reverse_dependencies,
-                    primary_packages,
-                    marker=markers.get(key_from_ireq(ireq)),
-                    hashes=hashes,
-                )
+                ireq_key = key_from_ireq(ireq)
```
Comment (webknjaz) [inline piptools/writer.py]: could as well generate this above as
```python
unsafe_requirements_map = ((key_from_ireq(r), r) for r in unsafe_requirements)
```
and iterate over that

File: piptools/writer.py [hunk 4]
```diff
@@ -189,17 +189,18 @@ def _iter_lines(
...
                 if not self.allow_unsafe:
-                    yield comment("# {}".format(req))
+                    yield comment("# {}".format(ireq_key))
```
Comment (webknjaz) [inline piptools/writer.py]: If you stick `continue` here, you could dedent the following code block
```suggestion
                    yield comment("# {}".format(ireq_key))
                    continue
```

File: tests/test_cli_compile.py (hunk 1 above)
Comment (webknjaz) [reply tests/test_cli_compile.py]: Oh, I see. The second one doesn't have `==`. It's been hard to spot. I suggest that you minimize the diff in order to make it review-friendly.

Comment (atugushev) [reply tests/test_cli_compile.py]: Yeah, that's why I've split the PR into two commits.

Comment (webknjaz) [reply tests/test_cli_compile.py]: Well, it's not visible in the PR diff and folks mostly don't do reviews per commit (I do sometimes but to all the time).

OTOH, you mix up two separate tasks in this PR:
1) reformatting the code
2) changing things

This makes it non-atomic, which is not good in my book. As you can see, this confuses even me. And I do a lot more reviews than most of the regular folks (compared to other types of contributions). Which opens up possibilities to sneak in a few more bugs unnoticed. Humans are error-prone and byte-by-byte comparisons are hard without any highlighting.

Comment (webknjaz) [inline tests/test_cli_compile.py]: Also, these two changes are even in the same commit.

Comment (atugushev) [reply tests/test_cli_compile.py]: What do you mean? This is the fix of the test.

Comment (atugushev) [reply tests/test_cli_compile.py]: Yeah, it should have been separate PRs. I thought these changes would be easy to review (which is apparently not). My bad. Would you like me to split this into two PRs or we could go on?

File: piptools/writer.py (hunk 3 above)
Comment (atugushev) [reply piptools/writer.py]: But we still need an `ireq` (which uses in `self._format_requirement`) reference in the for-loop body.

File: piptools/writer.py (hunk 2 above)
Comment (atugushev) [reply piptools/writer.py]: Yeah, we can do that. I did it the same way as it's done above (see 171th line). IMO the `yield line` looks more readable than `yield function call with many arguments` and it's debugger-friendly.

File: piptools/writer.py (hunk 4 above)
Comment (atugushev) [reply piptools/writer.py]: You probably right, but I think it's not related to the issue. I'd prefer to touch these lines in a separate PR. First things first.

Comment (atugushev) [reply piptools/writer.py]: And it looks subjective, is there a PEP which suggests to use:
```python
for x in z:
    if x:
        foo()
    else:
        bar()    
```

instead of

```python
for x in z:
    if x:
        foo()
        continue
    bar()    
```
? Just wondering.

File: tests/test_cli_compile.py (hunk 1 above)
Comment (webknjaz) [reply tests/test_cli_compile.py]: Yes, it would be better to avoid reformatting things in PRs that target changing the behavior.

File: piptools/writer.py (hunk 2 above)
Comment (webknjaz) [reply piptools/writer.py]: ok

File: piptools/writer.py (hunk 3 above)
Comment (webknjaz) [reply piptools/writer.py]: yes, by using `for ireq_key, ireq in unsafe_requirements_map:`

Comment (atugushev) [reply piptools/writer.py]: Oh, that's tricky. I'd prefer it keep simple :)

File: piptools/writer.py (hunk 4 above)
Comment (webknjaz) [reply piptools/writer.py]: Yes, it's subjective. I don't recall this being mentioned in any PEPs.
1) It reduces the indentation which is better for readability
2) There's a similar pylint rule for `return`
   ```python
   def func():
       if cond1:
           return xxx
       else:  # pylint would emit a warning here
           return yyy
   ```
   I think it's reasonable to apply this rule to other similar structures.
3) Yes, it makes sense to do this in a separate refactoring PR to keep the diff small.

File: tests/test_cli_compile.py (hunk 1 above)
Comment (atugushev) [reply tests/test_cli_compile.py]: I must note the first commit is not about reformatting. It improves the test. The actual test was not complete enough to make changes.

Comment (webknjaz) [reply tests/test_cli_compile.py]: This relates to the previous chain of comments explaining that it's bad to mix formatting and logic changes together.

Comment (atugushev) [reply tests/test_cli_compile.py]: "Yes" means split or go on?

Comment (webknjaz) [reply tests/test_cli_compile.py]: Please split

File: piptools/writer.py (hunk 3 above)
Comment (webknjaz) [reply piptools/writer.py]: ok

File: tests/test_cli_compile.py (hunk 1 above)
Comment (webknjaz) [reply tests/test_cli_compile.py]: P.S. these comments refer to the second commit which both (1) reformats 2 separate lines into one **and** (2) changes what's needed for the test. Hence two atomic changes in a single commit.

Comment (atugushev) [reply tests/test_cli_compile.py]: We use black. Reformat is necessary, but not the intention.

Comment (webknjaz) [reply tests/test_cli_compile.py]: Damn. black again...

Comment (atugushev) [reply tests/test_cli_compile.py]: > Please split

okay 👍

Comment (atugushev) [reply tests/test_cli_compile.py]: First one #986.

Comment (atugushev) [reply tests/test_cli_compile.py]: I've rebased the branch onto the master.

File: tests/test_cli_compile.py [hunk 5]
```diff
@@ -691,8 +691,7 @@ def test_annotate_option(pip_conf, runner, option, expected):
 
 
 @pytest.mark.parametrize(
-    "option, expected",
-    [("--allow-unsafe", "\nsetuptools=="), (None, "\n# setuptools==")],
```
Comment (atugushev) [inline tests/test_cli_compile.py]: FTR, would be nice to pin `setuptools` here as well, but it would take to rewrite the test.

File: tests/test_cli_compile.py (hunk 1 above)
Comment (atugushev) [inline tests/test_cli_compile.py]: FTR, would be nice to pin setuptools here as well, but it would take to rewrite the test.

Comment (webknjaz) [reply tests/test_cli_compile.py]: Wait, it still has reformatting and a useful change smashed in together.

Comment (atugushev) [reply tests/test_cli_compile.py]: Because first, we do useful change, then black reformats it. And we can't split it, because of the linter.

Comment (webknjaz) [reply tests/test_cli_compile.py]: And that's bad because the change doesn't show up highlighted in diff and it looks like you just took two lines and merged them into one because of this making the review next to impossible.

Comment (atugushev) [reply tests/test_cli_compile.py]: > you just took two lines and merged them into

this is what black does, not me :)

Comment (webknjaz) [reply tests/test_cli_compile.py]: you could always `# black off` to work around black's downsides

Comment (webknjaz) [reply tests/test_cli_compile.py]: I mean, if this keeps happening, we need to know how to make things review-friendly. Otherwise, we'll just silently accumulate bugs.

Comment (atugushev) [reply tests/test_cli_compile.py]: This is definitely what I don't want to do. Better to not use black at al then.

Comment (atugushev) [reply tests/test_cli_compile.py]: It would be nice if github's diff-tool would highlight changed string (which is removed "==" in this particular case).

Comment (atugushev) [reply tests/test_cli_compile.py]: > I mean, if this keeps happening, we need to know how to make things review-friendly. Otherwise, we'll just silently accumulate bugs.

Just wondering is there any discussion about black and non-friendly-review changes.  If not, let's open the discussion in black's issue-tracker for a start?

Comment (webknjaz) [reply tests/test_cli_compile.py]: Yes, I usually advocate for not using black because of its problems and freedoms it steels but it's not my call to make here, I guess :)

Regarding diff. I don't think that `git diff` as well as the default `diff` tool is capable of detecting string merge *and* edit. To these tools, it just looks like removing one line and editing another.

OTOH if you look to this diff line https://github.com/jazzband/pip-tools/pull/975/files#diff-9960f0e0139a2d6000305af0c80dac37R194 you'll see a part of the string with highlighted _which is what our goal should be_.

I dream of a tool that works similar to black but only reformats things that violate PEP8 and not everything unrelated to the target change...

Comment (webknjaz) [reply tests/test_cli_compile.py]: > Just wondering is there any discussion about black and non-friendly-review changes. If not, let's open the discussion in black's issue-tracker for a start?

Not that I know off. But whenever I start it, everybody starts hating me for this. So I'm not very enthusiastic about participating in yet another argument personally...

Comment (webknjaz) [reply tests/test_cli_compile.py]: I've converted this into an issue: #988.

Comment (atugushev) [reply tests/test_cli_compile.py]: Thanks!

Comment (atugushev) [reply tests/test_cli_compile.py]: > OTOH if you look to this diff line https://github.com/jazzband/pip-tools/pull/975/files#diff-9960f0e0139a2d6000305af0c80dac37R194 you'll see a part of the string with highlighted _which is what our goal should be_.

That's what I'm talking about!

> Not that I know of. But whenever I start it, everybody starts hating me for this. So I'm not very enthusiastic about participating in yet another argument personally...

Haha! That's a rough path 😄

Comment (webknjaz) [reply tests/test_cli_compile.py]: Anyway. I think that one commit should have a logic change and then autoformatting should be a separate commit.
Or just merge it if you want.

Comment (atugushev) [reply tests/test_cli_compile.py]: That makes sense, but as you mentioned before:
> Well, it's not visible in the PR diff and folks mostly don't do reviews per commit (I do sometimes but to all the time).
 
And I can't merge this unless someone approves it.

Comment (webknjaz) [reply tests/test_cli_compile.py]: Oh, that approval from @PeterJCLaw doesn't work..

Comment (webknjaz) [reply tests/test_cli_compile.py]: > as you mentioned before

I'd like to point out that both individual commits and PRs can be atomic, they just refer to different abstraction levels.

**Comment by atugushev** [comment]
@webknjaz thanks for reviewing this!
